### PR TITLE
Corda responder validation blog post

### PR DESCRIPTION
**Pull request by lankydan**

### CODE REVIEW

File: content/blog/2019/corda-responder-validation/index.md [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+published: true
+tags: [corda, kotlin, dlt, distributed ledger technology, blockchain]
+cover_image: ./title-card.png
+---
+
+Do you want to prevent your node from accepting transactions that are not important to your business?
+
+Do you want to prevent your node from accepting transactions that are invalid from your business's perspective?
+
+If you answered yes to either of these questions. Then you might want to consider adding some validation to your responder flows.
```
Comment (edward-prosser) [inline content/blog/2019/corda-responder-validation/index.md]: Change the full stop to a comma.

File: content/blog/2019/corda-responder-validation/index.md [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+
+For context, a responder flow is executed by a counterparty node when communicating with an initiating flow. Part of a responder flow's duty typically includes signing and recording a transaction sent to it. Meaning an organisation can end up in a situation where they did not have a large involvement in putting the contents of the transaction together, yet are expected to sign and potentially record it.
```
Comment (edward-prosser) [inline content/blog/2019/corda-responder-validation/index.md]: Too much qualification in this sentence: "_Part_ of a responder flow's duty _typically_..." Consider simplification: "Responder flows typically sign and record transactions sent to them."

Comment (edward-prosser) [reply content/blog/2019/corda-responder-validation/index.md]: Further, beginning the next sentence with "Meaning" is inelegant, consider "This means", or "Consequently"

File: content/blog/2019/corda-responder-validation/index.md [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+
+Surely, there should be some rules to prevent an organisation from being sent any old rubbish? And, yes, there are. The very bare minimum that needs to pass is contract validation. This ensures that a transaction is logically valid at a general level. What it does not do. Is enforce local identity/roles or business level requirements upon a transaction.
```
Comment (edward-prosser) [inline content/blog/2019/corda-responder-validation/index.md]: Why is there a full stop after "What it does not do"

Comment (edward-prosser) [reply content/blog/2019/corda-responder-validation/index.md]: Also, delete "upon a transaction", it's irrelevant in context.

File: content/blog/2019/corda-responder-validation/index.md [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+
+To an extent, this is fine. A __contract should__ only be ensuring that a transaction is __logically valid__.
```
Comment (edward-prosser) [inline content/blog/2019/corda-responder-validation/index.md]: Move the emphasis off contract, keep it on should, or remove it from both, either is fine.

File: content/blog/2019/corda-responder-validation/index.md [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+
+__Flows__ on the other hand, need to take charge in __enforcing business level requirements__ on transactions.
```
Comment (edward-prosser) [inline content/blog/2019/corda-responder-validation/index.md]: Change in -> of

File: content/blog/2019/corda-responder-validation/index.md [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+
+## Including validation
+
+To include your own validation, you need to make use of `SignTransactionFlow.checkTransaction` when calling the `SignTransactionFlow` subflow. I guarantee that you have seen `checkTransaction` before (if you have worked with Corda already). If you have not, then please tell me how you managed to do so.
```
Comment (edward-prosser) [inline content/blog/2019/corda-responder-validation/index.md]: Delete "you managed to do so."

File: content/blog/2019/corda-responder-validation/index.md [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+  check(!message.contents.containsSwearWords()) { "Mind your language" }
+  check(!message.contents.containsMemes()) { "Only serious messages are accepted" }
+  check(message.sender.name.organisation != "Nigerian Prince") { "Spam message detected" }
+  ```
+
+  I highly doubt that checking whether a message contains swear words would be contained inside of a contract. A message is still generally valid if it includes them. But, for the written flow, maybe it is a polite flow. Where any mention of a rude word must be beeped out (just like on pre-watershed TV). The same general gist applies to the other two rules.
+
+## How to validate
+
+I have just shown you what sort of validation to include. What I did not do is explain how the code shown before leads to a transaction being rejected. You might have sussed it out, but let me be thorough and go through it properly.
```
Comment (edward-prosser) [inline content/blog/2019/corda-responder-validation/index.md]: Changing "did not "-> "didn't" makes the sentence flow a little better in my opinion. Doesn't make a huge difference though.

File: content/blog/2019/corda-responder-validation/index.md [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+
+  ```kotlin
+  check(message.recipient == ourIdentity) { "I think you got the wrong person" }
+  ```
+
+Which one you throw depends on the context (and possibly personal preference). I also don't want to explain the difference between an `IllegalArgumentException` and `IllegalStateException` (as I would struggle). I personally prefer to use `check` here, but that's just my opinion.
+
+Both of these also have alternatives for null checking, `requireNotNull` and `checkNotNull`.
+
+Corda also provides a built-in way to do similar validation using the `reqiureThat` DSL (Domain Specific Language):
```
Comment (edward-prosser) [inline content/blog/2019/corda-responder-validation/index.md]: Spelling of `requireThat`

Comment (lankydan) [reply content/blog/2019/corda-responder-validation/index.md]: LOL 🤦‍♂

File: content/blog/2019/corda-responder-validation/index.md [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+  "I think you got the wrong person" using (message.recipient == ourIdentity)
+  // Non `using` lines can also be included
+  val contents = message.contents
+  "Mind your language" using (!contents.containsSwearWords())
+  "Only serious messages are accepted" using (!contents.containsMemes())
+  "Spam message detected" using (message.sender.name.organisation != "Nigerian Prince")
+}
+```
+
+`requireThat` was built for use inside of contracts but can be used anywhere. As the name suggests, it follows the same semantics as `require`. The difference is that it is a DSL, which makes use of `using` to define a message and criteria. Note, that you can add general code into the `requireThat` block. Only the `using` lines are going to (potentially) throw exceptions.
```
Comment (edward-prosser) [inline content/blog/2019/corda-responder-validation/index.md]: remove 'of'

File: content/blog/2019/corda-responder-validation/index.md [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+- __Business logic__
+  - Does that state have an acceptable value for my business? You should validate that the states you are receiving meet whatever criteria your business has. Contracts should not enforce business rules. This is why flows should perform business validation.
+
+## Closing thoughts
+
+I wrote this post to ensure that CorDapp developers add thorough validation to their applications. Ensuring that what they intend their CorDapps to do, is what they actually do. If there is even just a single hole in the hull of a ship, it is not a matter of whether it will sink, but when. Your applications should be put together with thorough consideration to reduce the chance (like a ship) that it sinks. Failures will still happen, but the chances of them happening and the repercussions that they cause will be significantly lower if your application is well defined and constrained. The validation inside of flows plays an essential part in fortifying the hull of your CorDapp.
```
Comment (edward-prosser) [inline content/blog/2019/corda-responder-validation/index.md:175]: Good analogy

Comment (lankydan) [reply content/blog/2019/corda-responder-validation/index.md:175]: 😊

**Pull request merged**